airflow/www/views.py

```diff
@@ -2676,6 +2676,138 @@ def task_instances(self):
 
         return json.dumps(task_instances, cls=utils_json.AirflowJsonEncoder)
 
+    @expose('/object/tree_data')
+    @auth.has_access(
+        [
+            (permissions.ACTION_CAN_READ, permissions.RESOURCE_DAG),
+            (permissions.ACTION_CAN_READ, permissions.RESOURCE_TASK_INSTANCE),
+        ]
+    )
+    @action_logging
+    def tree_data(self):
+        """Returns tree data"""
+        dag_id = request.args.get('dag_id')
+        dag = current_app.dag_bag.get_dag(dag_id)
+
+        if not dag:
+            flash(f'DAG "{dag_id}" seems to be missing from DagBag.', "error")
+            return redirect(url_for('Airflow.index'))
```

Comment:
Given this returns JSON, a 404 is probably more approriate

```suggestion
        if not dag:
            response = jsonify(
                {
                    'error': f"can't find dag {dag_id}"
                }
            )
            response.status_code = 404
            return response

```